devito/ir/iet/visitors.py

```diff
@@ -27,105 +27,7 @@
-               pass
-    """
-
-    def __init__(self):
-        handlers = {}
-        # visit methods are spelt visit_Foo.
-        prefix = "visit_"
-        # Inspect the methods on this instance to find out which
-        # handlers are defined.
-        for (name, meth) in inspect.getmembers(self, predicate=inspect.ismethod):
-            if not name.startswith(prefix):
-                continue
-            # Check the argument specification
-            # Valid options are:
-            #    visit_Foo(self, o, [*args, **kwargs])
-            argspec = inspect.getargspec(meth)
-            if len(argspec.args) < 2:
-                raise RuntimeError("Visit method signature must be "
-                                   "visit_Foo(self, o, [*args, **kwargs])")
-            handlers[name[len(prefix):]] = meth
-        self._handlers = handlers
-
-    """
-    :attr:`default_args`. A dict of default keyword arguments for the visitor.
-    These are not used by default in :meth:`visit`, however, a caller may pass
-    them explicitly to :meth:`visit` by accessing :attr:`default_args`.
-    For example::
-
-        .. code-block::
-           v = FooVisitor()
-           v.visit(node, **v.default_args)
-    """
-    default_args = {}
-
-    @classmethod
-    def default_retval(cls):
-        """A method that returns an object to use to populate return values.
-
-        If your visitor combines values in a tree-walk, it may be useful to
-        provide a object to combine the results into. :meth:`default_retval`
-        may be defined by the visitor to be called to provide an empty object
-        of appropriate type.
-        """
-        return None
-
-    def lookup_method(self, instance):
-        """Look up a handler method for a visitee.
-
-        :param instance: The instance to look up a method for.
-        """
-        cls = instance.__class__
-        try:
-            # Do we have a method handler defined for this type name
-            return self._handlers[cls.__name__]
-        except KeyError:
-            # No, walk the MRO.
-            for klass in cls.mro()[1:]:
-                entry = self._handlers.get(klass.__name__)
-                if entry:
-                    # Save it on this type name for faster lookup next time
-                    self._handlers[cls.__name__] = entry
-                    return entry
-        raise RuntimeError("No handler found for class %s", cls.__name__)
-
-    def visit(self, o, *args, **kwargs):
-        """Apply this :class:`Visitor` to an AST.
-
-            :param o: The :class:`Node` to visit.
-            :param args: Optional arguments to pass to the visit methods.
-            :param kwargs: Optional keyword arguments to pass to the visit methods.
-        """
-        meth = self.lookup_method(o)
-        return meth(o, *args, **kwargs)
-
-    def visit_object(self, o, **kwargs):
-        return self.default_retval()
+class Visitor(GenericVisitor):
```

Comment:
AFAIU those methods assume that the visitor is operating on an Iteration-Expression tree - which is not true for the visitors in the arguments module. Since the arguments visitors aren't using these methods anyway (and maybe they should - that might be causing some of the complexity in the visitors), I don't see why.

Reply:
oh yes you're right, forget this